MDC2020 compression

AndrewEdmonds11 opened the pull request:
Hi Everyone,

I've added a new module CompressDetStepMCs to do the step compression for MDC2020. I've also added it to the MDC2020 scripts. I've run low stats validations and am now running some higher stats to double-check everything. A couple of other small changes are that compression modules have now moved to a new Compression directory, and fhicl validation has been added to CompressDigiMCs.

I hope to present this in a Comp-Soft meeting soon but in the meantime let me know if you have any comments or questions.

Thanks,
Andy

FNALbuild commented:
Hi @AndrewEdmonds11,
You have proposed changes to files in these packages:
- JobConfig
- TrkHitReco
- Compression
- Filters

which require these tests: build.

@Mu2e/fnalbuild-users, @Mu2e/write have access to CI actions on [master](/Mu2e/Offline/tree/master).


:hourglass: The following tests have been triggered for 4ec4883667a0d44df7ecffe6eddd61447b0a38ee: build  (Build queue has 2 jobs)


[About FNALbuild.](https://mu2ewiki.fnal.gov/wiki/Git#GitHub_Pull_Request_Procedures_and_FNALbuild) [Code review on Mu2e/Offline.](https://mu2ewiki.fnal.gov/wiki/GitHubWorkflow#Code_Review)

FNALbuild commented:
:umbrella: The tests failed for ref 4ec4883667a0d44df7ecffe6eddd61447b0a38ee.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| merge | :heavy_check_mark: | Merged 4ec4883667a0d44df7ecffe6eddd61447b0a38ee at 515b455f9fe38adb0e344ed8fd976c6121e1d578 |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//359/artifact/scons.log). Build time: 13 min 36 sec |
| ceSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//359/artifact/ceSimReco.log) |
| transportOnly | :x: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//359/artifact/transportOnly.log) |
| PS | :x: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//359/artifact/PS.log) |
| g4study | :x: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//359/artifact/g4study.log) |
| cosmicSimReco | :x: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//359/artifact/cosmicSimReco.log) |
| rootOverlaps | :x: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//359/artifact/rootOverlaps.log) |
| g4surfaceCheck | :x: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//359/artifact/g4surfaceCheck.log) |
| g4test_03MT | :x: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//359/artifact/g4test_03MT.log) |
| FIXME, TODO count | :heavy_check_mark: | [TODO (0) FIXME (0) in 7 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//359/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [1 errors 0 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//359/artifact/clang-tidy.log) |


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//359/console).
Build artefacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.


Logfiles may also be accessed [at this link.](https://gist.github.com/16e6d71a52a3c74c01c02c0a71807f9c)

AndrewEdmonds11 commented:
It looks like I've broken some fhicl when I moved the compression modules. I will fix and then re test

brownd1978 commented:
Hi Andy,
JobConfig will also need updating.  But maybe you got that already?
Dave


On Tue, Jan 26, 2021 at 12:20 Andrew Edmonds <notifications@github.com>
wrote:

> It looks like I've broken some fhicl when I moved the compression modules.
> I will fix and then re test
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/324#issuecomment-767803194>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH575SOOBHCTQATMPKZDLS34PY3ANCNFSM4WUANWQQ>
> .
>
-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

AndrewEdmonds11 commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for 60abd32ce2d1c66336fa84335e0b4733d73c780c: build  (Build queue has 2 jobs)

AndrewEdmonds11 commented:
> 
> Hi Andy, JobConfig will also need updating. But maybe you got that already? Dave
> On Tue, Jan 26, 2021 at 12:20 Andrew Edmonds ***@***.***> wrote: It looks like I've broken some fhicl when I moved the compression modules. I will fix and then re test — You are receiving this because you are on a team that was mentioned. Reply to this email directly, view it on GitHub <[#324 (comment)](https://github.com/Mu2e/Offline/pull/324#issuecomment-767803194)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABAH575SOOBHCTQATMPKZDLS34PY3ANCNFSM4WUANWQQ> .
> -- David Nathan Brown Dave_Brown@lbl.gov Office Phone (510) 486-7261 Fax 495-2957 Lawrence Berkeley National Lab MS 50R5008 (50-6026C) Berkeley, CA 94720

Hi Dave,

I've made the JobConfig changes already (i.e. I added compressDetStepMCs to JobConfig/primary/prolog.fcl). If there's anything I've missed, then let me know :)

Cheers,
Andy

FNALbuild commented:
:sunny: The tests passed at ref 60abd32ce2d1c66336fa84335e0b4733d73c780c.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| merge | :heavy_check_mark: | Merged 60abd32ce2d1c66336fa84335e0b4733d73c780c at 515b455f9fe38adb0e344ed8fd976c6121e1d578 |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//360/artifact/scons.log). Build time: 14 min 20 sec |
| ceSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//360/artifact/ceSimReco.log) |
| transportOnly | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//360/artifact/transportOnly.log) |
| PS | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//360/artifact/PS.log) |
| g4study | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//360/artifact/g4study.log) |
| cosmicSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//360/artifact/cosmicSimReco.log) |
| rootOverlaps | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//360/artifact/rootOverlaps.log) |
| g4surfaceCheck | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//360/artifact/g4surfaceCheck.log) |
| g4test_03MT | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//360/artifact/g4test_03MT.log) |
| FIXME, TODO count | :heavy_check_mark: | [TODO (0) FIXME (0) in 7 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//360/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [1 errors 0 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//360/artifact/clang-tidy.log) |


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//360/console).
Build artefacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.


Logfiles may also be accessed [at this link.](https://gist.github.com/fe51806968c6c119e072f428260ecf22)

brownd1978 commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for 4ee4463a45a30dfa138d173cf1ef5246e8afd802: build  (Build queue is empty)

brownd1978 commented on Compression/fcl/prolog.fcl:
```diff
@@ -0,0 +1,204 @@
+BEGIN_PROLOG
+
```
This prolog creates circular dependencies with JobConfig, in that JobConfig must source this prolog, but this prolog uses names set by JobConfig.  I know this is how we did this for MDC2018, but I now prefer a one-directional config structure.  I suggest moving most of this content into the appropriate JobConfig prolog (primary or pileup).

AndrewEdmonds11 replied:
I've reoriganised the prolog in commit 57645ffe4

brownd1978 commented on Compression/fcl/prolog.fcl:
```diff
@@ -0,0 +1,204 @@
+  compressDetStepMCs_none : {
+    strawGasStepCompressionLevel : "noCompression" # keep all StrawGasSteps
+    caloShowerStepCompressionLevel : "noCompression"
+    crvStepCompressionLevel : "noCompression"
```
Could you configure noCompression automatically, ie if the input tag is set to ""?  That would simplify the scripts.

AndrewEdmonds11 replied:
I think there's a difference between noCompression and not having the input collection. For the noCompression option, the idea is to just copy the input collection to an output collection so that it has the correct module label. This way, when debugging, we don't need to change downstream fhicl. 

Your comment does raise the question about what this module will do if we don't create one of the step collections. At the moment, it will throw an exception but maybe we want the flexibility to run with just only some of the detector steps?

FNALbuild commented:
:sunny: The tests passed at ref 4ee4463a45a30dfa138d173cf1ef5246e8afd802.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| merge | :heavy_check_mark: | Merged 4ee4463a45a30dfa138d173cf1ef5246e8afd802 at 33a8f519cac4e5b7a55704dca696549d44cca752 |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//366/artifact/scons.log). Build time: 13 min 14 sec |
| ceSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//366/artifact/ceSimReco.log) |
| transportOnly | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//366/artifact/transportOnly.log) |
| PS | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//366/artifact/PS.log) |
| g4study | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//366/artifact/g4study.log) |
| cosmicSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//366/artifact/cosmicSimReco.log) |
| rootOverlaps | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//366/artifact/rootOverlaps.log) |
| g4surfaceCheck | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//366/artifact/g4surfaceCheck.log) |
| g4test_03MT | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//366/artifact/g4test_03MT.log) |
| FIXME, TODO count | :heavy_check_mark: | [TODO (0) FIXME (0) in 7 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//366/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [1 errors 0 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//366/artifact/clang-tidy.log) |


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//366/console).
Build artefacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.


Logfiles may also be accessed [at this link.](https://gist.github.com/4c3452b679cd055090eea5cca4159f63)

brownd1978 commented on Compression/src/CompressDetStepMCs_module.cc:
```diff
@@ -0,0 +1,638 @@
+    using Comment=fhicl::Comment;
+
+    // fhicl parameters for all of our inputs
+    fhicl::Atom<art::InputTag> strawGasStepTag{Name("strawGasStepTag"), Comment("InputTag for the StrawGasSteps")};
```
For mixing we will have multiple Step collections.  I think the step input config needs to be a Sequence, not an Atom.

AndrewEdmonds11 replied:
I don't see us running this module when mixing unless I've missed something. Is the idea to create datasets with primary+pileup steps? I thought we would be taking primary steps and pileup steps to create primary+pileup digis.

brownd1978 commented on Compression/src/CompressDetStepMCs_module.cc:
```diff
@@ -0,0 +1,638 @@
+    StrawGasStep newStrawGasStep(i_strawGasStep);
+    _newStrawGasSteps->push_back(newStrawGasStep);
+  }
+  if(_strawGasStepCompressionLevel == mu2e::CompressionLevel::kNoCompression) {
```
From looking at the code above,  newStrawGasSteps must always have the same length as strawGasSteps, so why is this test necessary?  And what is the use case for skipping the test (setting a higher compression level)?  Same comment for the other steps.

AndrewEdmonds11 replied:
This test is just a sanity check in case I add code that inadvertantly remove steps. I don't see a use case for compressing steps in this module. I'm thinking in case we want to refactor CompressDigiMCs to have a similar structure, then there would be some consistency between different compression modules. I also like the idea of being explicit in the configuration about which data products we are touching.

brownd1978 commented on Compression/src/CompressDetStepMCs_module.cc:
```diff
@@ -0,0 +1,638 @@
+// - SimParticles : noCompression, fullCompression
+// - StepPointMCs : noCompression, simParticleCompression
+// - MCTrajectories : noCompression, simParticleCompression
+//
```
You need to add MCPrimary here too.  That (generally) gets produced in the same job as runs the step compression.

AndrewEdmonds11 replied:
Do you mean PrimaryParticle? I did have this in but removed it because there was a comment in the JobConfig scripts that said the FindMCPrimary module would need to be reconfigured for after compression. So I did that instead.

brownd1978 commented on Compression/src/CompressDetStepMCs_module.cc:
```diff
@@ -0,0 +1,638 @@
+          // If we haven't added a truncated particle in this genealogy
+          if (!addedTruncated) {
+            // Add a truncated particle
+            truncated = new SimParticle();
```
It looks like a single truncated particle is shared by all truncated genealogies.  It also appears it has no content.  Did you consider creating a unique truncated particle for each genealogy, including some content (like the ultimate parent, or the ultimate gen particle)?

AndrewEdmonds11 replied:
There should be a truncated particle for each genealogy branch. The only information I've set in the truncated particle is the parent ptr. I thought about adding a unique process code so that users can identify these truncated particle. I will bring this up to discuss in my talk. Also, the ultimate parent is always kept so that information should be safe.

brownd1978 commented on Compression/src/CompressDetStepMCs_module.cc:
```diff
@@ -0,0 +1,638 @@
+  }
+}
+
+void mu2e::CompressDetStepMCs::checkCompressionLevels() {
```
I am confused why step compression level is an option when there is only 1 allowed value.  Do you have a use case in mind where we would actually remove steps?  I can't think of one.

AndrewEdmonds11 replied:
I don't see us removing steps in this module. I like to be explicit about what we're touching and what we're doing and, if we end up refactoring CompressDigiMCs to match this module more, it will provide some consistency.

brownd1978 commented on Compression/src/CompressDigiMCsCheck_module.cc:
```diff
@@ -21,6 +21,25 @@ using namespace std;
 
   class CompressDigiMCsCheck : public art::EDAnalyzer {
+
+    struct Config {
```
thanks a lot for updating this!

brownd1978 commented on Compression/src/CompressDigiMCsCheck_module.cc:
```diff
@@ -21,6 +21,25 @@ using namespace std;
+      fhicl::Atom<art::InputTag> oldCrvDigiMCTag{Name("oldCrvDigiMCTag"), Comment("InputTag for uncompressed CrvDigiMCCollection")};
+      fhicl::Atom<art::InputTag> newCrvDigiMCTag{Name("newCrvDigiMCTag"), Comment("InputTag for compressed CrvDigiMCCollection")};
+      fhicl::Atom<art::InputTag> crvDigiMCIndexMapTag{Name("crvDigiMCIndexMapTag"), Comment("If a CrvDigiMCIndexMap was passed to the compression originally, then pass it here too")};
+      fhicl::Atom<art::InputTag> oldCaloShowerSimTag{Name("oldCaloShowerSimTag"), Comment("InputTag for uncompressed CaloShowerSimCollection")};
```
I don't think we will ever compress ShowerSim given the new workflows.  It doesn't hurt to keep it though.

brownd1978 commented on Compression/src/CompressDigiMCsCheck_module.cc:
```diff
@@ -239,15 +258,15 @@ namespace mu2e {
+          throw cet::exception("CompressDigiMCsCheck") << "Old and new CaloShowerStepPtrs do not match" << std::endl;
+        }
       }
     }
```
There is no check of CaloClusterMC.  That is used in the current workflow.  I suggest to replace the CaloShowerSim check with a CaloClusterMC check.

AndrewEdmonds11 replied:
Added in commit 337430fac

brownd1978 commented on Compression/src/CompressDigiMCs_module.cc:
```diff
@@ -89,7 +89,39 @@ namespace mu2e {
+    using Comment=fhicl::Comment;
+
+    // fhicl parameters for all of our inputs
+    fhicl::Atom<art::InputTag> strawDigiMCTag{Name("strawDigiMCTag"), Comment("InputTag for the StrawDigiMCCollection")};
```
thanks for updating this!

brownd1978 commented on Compression/src/CompressDigiMCs_module.cc:
```diff
@@ -119,16 +151,24 @@ class mu2e::CompressDigiMCs : public art::EDProducer {
+  art::InputTag _crvDigiMCIndexMapTag;
   std::vector<art::InputTag> _caloShowerStepTags;
   art::InputTag _caloShowerSimTag;
-  art::InputTag _CaloShowerROTag;
```
same comment as before about CaloShowerSim.  Also CaloShowerRO isn't kept so doesn't need compression in our new workflows.

brownd1978 reviewed (changes requested):
This PR provides an important new technology (step compression), and an important update to existing technology (digi compression).  There are a few parts that need modification, but the general architecture and implementation look good.

brownd1978 commented:
Yes you are right.  We want to compress before mixing.  Please ignore that
comment.

On Wed, Jan 27, 2021 at 10:06 AM Andrew Edmonds <notifications@github.com>
wrote:

> *@AndrewEdmonds11* commented on this pull request.
> ------------------------------
>
> In Compression/src/CompressDetStepMCs_module.cc
> <https://github.com/Mu2e/Offline/pull/324#discussion_r565523108>:
>
> > +    std::set<cet::map_vector_key> m_keys;
> +
> +  };
> +
> +  typedef std::string InstanceLabel;
> +}
> +
> +
> +class mu2e::CompressDetStepMCs : public art::EDProducer {
> +public:
> +  struct Config {
> +    using Name=fhicl::Name;
> +    using Comment=fhicl::Comment;
> +
> +    // fhicl parameters for all of our inputs
> +    fhicl::Atom<art::InputTag> strawGasStepTag{Name("strawGasStepTag"), Comment("InputTag for the StrawGasSteps")};
>
> I don't see us running this module when mixing unless I've missed
> something. Is the idea to create datasets with primary+pileup steps? I
> thought we would be taking primary steps and pileup steps to create
> primary+pileup digis.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/324#discussion_r565523108>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH573QXXR3KA3RXYUVPQDS4BIZ5ANCNFSM4WUANWQQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

brownd1978 commented:
OK, that's a better solution.

On Wed, Jan 27, 2021 at 10:11 AM Andrew Edmonds <notifications@github.com>
wrote:

> *@AndrewEdmonds11* commented on this pull request.
> ------------------------------
>
> In Compression/src/CompressDetStepMCs_module.cc
> <https://github.com/Mu2e/Offline/pull/324#discussion_r565526703>:
>
> > @@ -0,0 +1,638 @@
> +//
> +// Compression module that takes detector steps (StrawGasSteps, CaloShowerSteps, and CrvSteps)
> +// and compresses out unwanted MC information (StepPointMCs, SimParticles and GenParticles).
> +// Note that the number of detector steps is NOT reduced.
> +//
> +// Allowed compression levels for different data products:
> +// - StrawGasSteps : noCompression
> +// - CaloShowerSteps : noCompression
> +// - CrvSteps : noCompression
> +// - SimParticles : noCompression, fullCompression
> +// - StepPointMCs : noCompression, simParticleCompression
> +// - MCTrajectories : noCompression, simParticleCompression
> +//
>
> Do you mean PrimaryParticle? I did have this in but removed it because
> there was a comment in the JobConfig scripts that said the FindMCPrimary
> module would need to be reconfigured for after compression. So I did that
> instead.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/324#discussion_r565526703>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH57324LLWSDZAUFAIAALS4BJOFANCNFSM4WUANWQQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

AndrewEdmonds11 commented:
Thanks for the review, Dave. I've replied to everything that I think might need further discussion and I agree with everything else so will make those changes soon.

brownd1978 commented:
I see.  so you let downstream users know the collection was rewritten, but
only allow 'nocompression' as level.  Since we have no use case for
anything except nocompression, and the code throws withy anything else,
I suggest to make this the config default.  That will avoid cluttering the
fcl files.
As for steps, going forwards we will always require Straw, Calo and Crv
step collections in every production file, though some may be empty.

On Wed, Jan 27, 2021 at 10:10 AM Andrew Edmonds <notifications@github.com>
wrote:

> *@AndrewEdmonds11* commented on this pull request.
> ------------------------------
>
> In Compression/src/CompressDetStepMCs_module.cc
> <https://github.com/Mu2e/Offline/pull/324#discussion_r565525900>:
>
> > +
> +void mu2e::CompressDetStepMCs::compressStrawGasSteps(const art::Event& event) {
> +  art::Handle<mu2e::StrawGasStepCollection> strawGasStepsHandle;
> +  event.getByLabel(_conf.strawGasStepTag(), strawGasStepsHandle);
> +  const auto& strawGasSteps = *strawGasStepsHandle;
> +  if(_conf.debugLevel()>0 && strawGasSteps.size()>0) {
> +    std::cout << "Compressing StrawGasSteps from " << _conf.strawGasStepTag() << std::endl;
> +  }
> +  for (const auto& i_strawGasStep : strawGasSteps) {
> +    if (_simParticleCompressionLevel == mu2e::CompressionLevel::kFullCompression) {
> +      recordSimParticle(i_strawGasStep.simParticle());
> +    }
> +    StrawGasStep newStrawGasStep(i_strawGasStep);
> +    _newStrawGasSteps->push_back(newStrawGasStep);
> +  }
> +  if(_strawGasStepCompressionLevel == mu2e::CompressionLevel::kNoCompression) {
>
> This test is just a sanity check in case I add code that inadvertantly
> remove steps. I don't see a use case for compressing steps in this module.
> I'm thinking in case we want to refactor CompressDigiMCs to have a similar
> structure, then there would be some consistency between different
> compression modules. I also like the idea of being explicit in the
> configuration about which data products we are touching.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/324#discussion_r565525900>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH57ZCJGLUW52TT3BYPN3S4BJJLANCNFSM4WUANWQQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

brownd1978 commented:
So the position and momentum members of this SimParticle are meaningless?
In that case, I think a dedicated process code is essential.  Longer-term,
we may wish to add a layer to the SimParticle inheritance so the truncated
particle doesn't have to carry invalid payload.
Good to know about the ultimate parent, that makes sense.  I think a use
case (diag module) showing how truncated particles can be used would be
helpful.

On Wed, Jan 27, 2021 at 10:18 AM Andrew Edmonds <notifications@github.com>
wrote:

> *@AndrewEdmonds11* commented on this pull request.
> ------------------------------
>
> In Compression/src/CompressDetStepMCs_module.cc
> <https://github.com/Mu2e/Offline/pull/324#discussion_r565531003>:
>
> > +      cet::map_vector_key truncKey;
> +      art::Ptr<mu2e::SimParticle> truncSimPtr;
> +      SimParticle* truncated = 0;
> +
> +      art::Ptr<mu2e::SimParticle> i_childPtr = i_keptSimPart;//art::Ptr<mu2e::SimParticle>(_newSimParticlesPID, i_newSimParticle.first.asUint(), _newSimParticleGetter);
> +      art::Ptr<mu2e::SimParticle> i_parentPtr = i_childPtr->parent();
> +      while (i_parentPtr) {
> +        if (_simPtrRemap.find(i_parentPtr) == _simPtrRemap.end()) { // the parent is not in the output collection
> +          if (_conf.debugLevel()>0) {
> +            std::cout << "SimParticle " << i_parentPtr << " is not in output collection because it has been compressed away by genealogy compression" << std::endl;
> +          }
> +
> +          // If we haven't added a truncated particle in this genealogy
> +          if (!addedTruncated) {
> +            // Add a truncated particle
> +            truncated = new SimParticle();
>
> There should be a truncated particle for each genealogy branch. The only
> information I've set in the truncated particle is the parent ptr. I thought
> about adding a unique process code so that users can identify these
> truncated particle. I will bring this up to discuss in my talk. Also, the
> ultimate parent is always kept so that information should be safe.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/324#discussion_r565531003>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH572AMMYHF6F7X2P5FEDS4BKGVANCNFSM4WUANWQQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

brownd1978 commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for f9e60ddd4211549180dbb0d2adbc8e0135d3d14d: build  (Build queue has 3 jobs)

brownd1978 commented on JobConfig/primary/prolog.fcl:
```diff
@@ -21,7 +21,18 @@ Primary: {
+      stepPointMCTags : [ "g4run:virtualdetector", "g4run:protonabsorber" ]#, "g4run:stoppingtarget" ]
+      mcTrajectoryTag : "g4run"
+      debugLevel : 0
+      compressionOptions : @local::DetStepCompression.standardCompression
```
This is a nice factorization

brownd1978 commented on JobConfig/reco/prolog.fcl:
```diff
@@ -130,14 +130,15 @@ Reconstruction : {
       caloShowerROTag : ""
       caloClusterMCTag : "CaloClusterTruthMatch"
-      caloHitMCTag : "CaloHitTruthMatch"
+#      caloHitMCTag : "CaloHitTruthMatch"
```
I'm not sure why CaloHitMCTag is commented out.  Is it because the step filter already does this?  I think these need to be re-written anyways, just because the pointers change with the CaloClusterMCs are compressed.

AndrewEdmonds11 replied:
The new validated fhicl was throwing an error because this fhicl parameter isn't defined. I thought about adding it but it looked like we don't use caloHitMCTag anywhere. I think we just get to CaloHitMCs through CaloClusterMCs but I meant to look into this further in case it was a rebase error on my part.

brownd1978 reviewed (commented):
Looks good Andy!  I just have one question about the reco compression config, after that this looks good to merge.

FNALbuild commented:
:sunny: The tests passed at ref f9e60ddd4211549180dbb0d2adbc8e0135d3d14d.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| merge | :heavy_check_mark: | Merged f9e60ddd4211549180dbb0d2adbc8e0135d3d14d at 54cf3799c7dcc92a462303bb425ab60f460c88f3 |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//373/artifact/scons.log). Build time: 15 min 00 sec |
| ceSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//373/artifact/ceSimReco.log) |
| transportOnly | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//373/artifact/transportOnly.log) |
| PS | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//373/artifact/PS.log) |
| g4study | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//373/artifact/g4study.log) |
| cosmicSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//373/artifact/cosmicSimReco.log) |
| rootOverlaps | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//373/artifact/rootOverlaps.log) |
| g4surfaceCheck | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//373/artifact/g4surfaceCheck.log) |
| g4test_03MT | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//373/artifact/g4test_03MT.log) |
| FIXME, TODO count | :heavy_check_mark: | [TODO (0) FIXME (0) in 12 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//373/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [1 errors 0 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//373/artifact/clang-tidy.log) |


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//373/console).
Build artefacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.


Logfiles may also be accessed [at this link.](https://gist.github.com/0a5ed7b2a2e70cb2c05232061f05b93c)

brownd1978 commented:
Yes this makes sense, we don't need this parameter.  Please just delete
that line.
cheers, Dave

On Tue, Feb 2, 2021 at 3:37 PM Andrew Edmonds <notifications@github.com>
wrote:

> *@AndrewEdmonds11* commented on this pull request.
> ------------------------------
>
> In JobConfig/reco/prolog.fcl
> <https://github.com/Mu2e/Offline/pull/324#discussion_r569005977>:
>
> > @@ -130,14 +130,15 @@ Reconstruction : {
>        caloShowerSimTag : ""
>        caloShowerROTag : ""
>        caloClusterMCTag : "CaloClusterTruthMatch"
> -      caloHitMCTag : "CaloHitTruthMatch"
> +#      caloHitMCTag : "CaloHitTruthMatch"
>
> The new validated fhicl was throwing an error because this fhicl parameter
> isn't defined. I thought about adding it but it looked like we don't use
> caloHitMCTag anywhere. I think we just get to CaloHitMCs through
> CaloClusterMCs but I meant to look into this further in case it was a
> rebase error on my part.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/324#discussion_r569005977>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH577V55NF4MNSGGDWRA3S5CECNANCNFSM4WUANWQQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

brownd1978 commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for f692eb334a41e5704d90b80258d9d69cbae979e2: build  (Build queue has 2 jobs)

FNALbuild commented:
:sunny: The tests passed at ref f692eb334a41e5704d90b80258d9d69cbae979e2.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| merge | :heavy_check_mark: | Merged f692eb334a41e5704d90b80258d9d69cbae979e2 at 1b5c92127b4756d44f329a739dba5c0f273ed0b0 |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//377/artifact/scons.log). Build time: 13 min 58 sec |
| ceSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//377/artifact/ceSimReco.log) |
| transportOnly | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//377/artifact/transportOnly.log) |
| PS | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//377/artifact/PS.log) |
| g4study | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//377/artifact/g4study.log) |
| cosmicSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//377/artifact/cosmicSimReco.log) |
| rootOverlaps | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//377/artifact/rootOverlaps.log) |
| g4surfaceCheck | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//377/artifact/g4surfaceCheck.log) |
| g4test_03MT | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//377/artifact/g4test_03MT.log) |
| FIXME, TODO count | :heavy_check_mark: | [TODO (0) FIXME (0) in 12 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//377/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [1 errors 0 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//377/artifact/clang-tidy.log) |


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//377/console).
Build artefacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.


Logfiles may also be accessed [at this link.](https://gist.github.com/ae523e0601c589edf8c10f736784d7be)

gaponenko commented on Compression/src/SConscript:
```diff
@@ -0,0 +1,74 @@
+    'art_Framework_Services_Registry',
+    'art_root_io_tfile_support',
+    'art_root_io_TFileService_service',
+    'art_Framework_Services_Optional_RandomNumberGenerator_service',
```
Why does compression depend on random numbers?  Please purge unneeded libraries from the list.

AndrewEdmonds11 replied:
I've purged what I can in commit 3abdea0

gaponenko reviewed (changes requested):
Hi Andy, nice work!   I made a few notes on the coding, please see below.
Andrei

gaponenko commented on Compression/src/CompressDetStepMCs_module.cc:
```diff
@@ -0,0 +1,671 @@
+}
+void mu2e::CompressDetStepMCs::compressMCTrajectories(const art::Event& event) {
+
+  const auto& mcTrajectories = event.getValidHandle<MCTrajectoryCollection>(_conf.mcTrajectoryTag());
```
I forgot to mention another important thing.   The compression modules must allow some of their input products to be ignored, by e.g. explicitly specifying the empty string for the tag in fhicl config.  Otherwise these large modules will not be usable outside of a narrowly defined "production" job configuration.  I am attaching this comment to a line where the code will break if my file does not have trajectories saved, but this comment is more general and applies to other products and the other module as well.

brownd1978 replied:
We are evolving to require common inputs for all jobs of a given type.  Fhicl is simply too limited to handle that internally, and the proliferation of scripts needed to deal inconsistent, incompatible inputs is unmaintainable.   What use case are you thinking of that requires this?

AndrewEdmonds11 replied:
I've added checks for empty tags as a way to do this in commit e59f3ca

gaponenko commented:
Hi,

> We are evolving to require common inputs for all jobs of a given type.

That's not relevant to my comment.  Of course we will make the module
to work for one hardwired configuration.

An example that comes to mind is the old mixing module, now deleted
from the repo.  It was impossible to use that module any data product
it knew about was not present.  It did not serve us well.

>   What use case are you thinking of that requires this?

I do not have an immediate use case now.  It does not mean we won't
get one tomorrow.  It is cheaper to make C++ code flexible now while
we are reviewing it.

Andrei

brownd1978 commented:
An empty sequence is a natural and unambiguous way to describe absence of
data, I use that frequently.  A null string however is different: it's a
value to which you want to ascribe a particular meaning.
I suspect the empty string will also break one of your earlier suggestions,
namely tying the 'consumes' statement directly to translating to a token; I
believe the code needs extra logic to deal with that.


On Mon, Feb 8, 2021 at 10:23 AM Andrei Gaponenko <notifications@github.com>
wrote:

> Hi,
>
> > We are evolving to require common inputs for all jobs of a given type.
>
> That's not relevant to my comment. Of course we will make the module
> to work for one hardwired configuration.
>
> An example that comes to mind is the old mixing module, now deleted
> from the repo. It was impossible to use that module any data product
> it knew about was not present. It did not serve us well.
>
> > What use case are you thinking of that requires this?
>
> I do not have an immediate use case now. It does not mean we won't
> get one tomorrow. It is cheaper to make C++ code flexible now while
> we are reviewing it.
>
> Andrei
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/324#issuecomment-775347147>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH575L243X5SN7C2DW7J3S6AT2DANCNFSM4WUANWQQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

gaponenko commented:
Hi Dave,

On Mon, 8 Feb 2021, David Brown wrote:

> An empty sequence is a natural and unambiguous way to describe absence of
> data, I use that frequently.  A null string however is different: it's a
> value to which you want to ascribe a particular meaning.

Yes.  And I think it is more practical to ascribe a meaning to "" than
to add a boolean "use" flag for every input tag.

> I suspect the empty string will also break one of your earlier suggestions,
> namely tying the 'consumes' statement directly to translating to a token; I
> believe the code needs extra logic to deal with that.

I think you are right. But this is a general "consumes" problem that
is not tied to the emtpy string interpretation.  Rather it is tied to
dealing with optional input collections.  I will ask art people about
how to handle this.  For now we can keep using InputTags and not
switch to tokens.

Andrei

brownd1978 commented:
> I suspect the empty string will also break one of your earlier suggestions,
> > namely tying the 'consumes' statement directly to translating to a token; I
> > believe the code needs extra logic to deal with that.
> 
> I think you are right. But this is a general "consumes" problem that
> is not tied to the emtpy string interpretation. Rather it is tied to
> dealing with optional input collections.

An empty sequence handles ‘consumes’ without extra logic or protection code or any change to art.  If input data are optional I recommend using a sequence. 

> I will ask art people about
> how to handle this. For now we can keep using InputTags and not
> switch to tokens.
> 
> Andrei
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

gaponenko commented:
> An empty sequence handles ‘consumes’ without extra logic or protection code or any change to art.  If input data are optional I recommend using a sequence. 

But then one needs a plan to deal with more than one input.  One would
need to declare multiple output products for those inputs.  And invent
a way to generate output instance names relating InputTag of each
collection to its output instance...  Way more complicated than
dealing with a single optional input.

Andrei

brownd1978 commented:
On Mon, Feb 8, 2021 at 12:51 PM Andrei Gaponenko <notifications@github.com>
wrote:

> > An empty sequence handles ‘consumes’ without extra logic or protection
> code or any change to art. If input data are optional I recommend using a
> sequence.
>
> But then one needs a plan to deal with more than one input. One would
> need to declare multiple output products for those inputs. And invent
> a way to generate output instance names relating InputTag of each
> collection to its output instance... Way more complicated than
> dealing with a single optional input.
>

... or not.  All inputs going to one output is perfectly fine and easy.

It is overkill in the case you know there's only ever going to be 1 input
collection, but I still prefer it to the extra logic (if (tag != "")) then.

Dave


>
> Andrei
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/324#issuecomment-775451819>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH57435DODH5QGU4RWVPDS6BFD5ANCNFSM4WUANWQQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

gaponenko commented:
> > But then one needs a plan to deal with more than one input. One would
> > need to declare multiple output products for those inputs. And invent
> > a way to generate output instance names relating InputTag of each
> > collection to its output instance... Way more complicated than
> > dealing with a single optional input.
> >
> 
> ... or not.  All inputs going to one output is perfectly fine and easy.
> 
> It is overkill in the case you know there's only ever going to be 1 input
> collection, but I still prefer it to the extra logic (if (tag != "")) then.

Either way is fine with me, as long as there is a way to run a job with 
some products missing.  Of course the "missing" case should  not be the 
default so that mistakes are still caught.

Andrei

gaponenko commented:
I have not heard any comments on the below.

If some years from now we need to add compression for e.g. RPC planes,  a 
student would be able to write simple modules to handle steps (1) and (3).
All the exiting functionality would be safe as modules are isolated.
As opposed to dealing with the large monolithic module and potentially
breaking old stuff while adding new one.

Andrei

On Mon, 8 Feb 2021, Andrei Gaponenko wrote:

> More on this:
> 
> On Mon, 8 Feb 2021, Andrei Gaponenko wrote:
> 
> > Hi Dave,
> > 
> > On Mon, 8 Feb 2021, David Brown wrote:
> > 
> > > An empty sequence is a natural and unambiguous way to describe absence of
> > > data, I use that frequently.  A null string however is different: it's a
> > > value to which you want to ascribe a particular meaning.
> 
> A cleaner approach would be to deal with the minimal number of data
> products per module and use several modules instead of using the
> monolith.   Like this:
> 
> 1) A module that looks at just StrawGasSteps and identifies which
>    SimParticles should be preserved.  Writes out a collection of
>    art::Ptr<SimParticle> to keep.
> 
>    Similar modules for Calo etc.
>    
>    FCL configuration runs those modules that are relevant to the job,
>    if CRV is missing just do not run its module.
> 
> 2) A module that looks at all the Ptrs and compresses the
>    SimParticleCollection, and a specified list of StepPointMC
>    collections based on kept particles.  It also produces a
>    SimParticleRemapping product.
> 
> 3) Per-detector "postprocessing" modules that rewrite their
>    steps collections to point to the compressed SimParticles,
>    using the remapping product.
> 
> This is the model I had in mind back when FilterG4Out module was
> written.  It provides support for step (2).  Back in 2013 there were
> no detector specific steps to implement anything else.  (I did not
> plan for SimParticle history truncation, so that's a difference from
> the new monolithic module functionality - but one can also implement
> SimParticle truncation in (2)).
> 
> Andrei
> 
> -- 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_324-23issuecomment-2D775428821&d=DwICaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=O47fc5vzDTR2V_gla4Ub0Q&m=U6OD3hxmnvL_yUgRq1qWB10PLQyew5YA-hjm5nG-PTE&s=3yrkRiay3a8t_6gegagSV6IOIif4P7FNyNOD8Xy2ZSI&e=

brownd1978 commented:
Step compression is a well-defined and self-contained process, it makes sense to put it into 1 module.  I don't see any advantage and there is clear cost in duplicated code, extra maintenance, more complicated configuration, extra module overhead, additional output payload bloat, etc. by splitting each input container processed into its own module.  The extension to hypothetical new detectors is no more work in this design than any other.

AndrewEdmonds11 commented on Compression/fcl/prolog.fcl:
```diff
@@ -0,0 +1,166 @@
+    simParticleCompressionLevel : "fullCompression" # only keep SimParticles related to other data products we are keeping
+    stepPointMCCompressionLevel : "simParticleCompression" # only keep StepPointMCs related to SimParticles we are keeping
+    keepNGenerations : -1
+    truncatedSimParticleKeyOffset : 400000
```
Yes, this is removed with the new genealogy compression in commit 81ef00e

AndrewEdmonds11 commented on Compression/src/CompressDetStepMCs_module.cc:
```diff
@@ -0,0 +1,671 @@
+  CompressDetStepMCs(CompressDetStepMCs const &) = delete;
+  CompressDetStepMCs(CompressDetStepMCs &&) = delete;
+  CompressDetStepMCs & operator = (CompressDetStepMCs const &) = delete;
+  CompressDetStepMCs & operator = (CompressDetStepMCs &&) = delete;
```
The perils of copying old modules. I've removed these lines in CompressDetStepMCs (commit 7b40442) and CompressDigiMCs (commit c26897c)

AndrewEdmonds11 commented on Compression/src/CompressDetStepMCs_module.cc:
```diff
@@ -0,0 +1,671 @@
+
+void mu2e::CompressDetStepMCs::compressStrawGasSteps(const art::Event& event) {
+  art::Handle<mu2e::StrawGasStepCollection> strawGasStepsHandle;
+  event.getByLabel(_conf.strawGasStepTag(), strawGasStepsHandle);
```
Moved to using product tokens in commit 523713c

AndrewEdmonds11 commented on Compression/src/CompressDetStepMCs_module.cc:
```diff
@@ -0,0 +1,671 @@
+    }
+    else if (_mcTrajectoryCompressionLevel == mu2e::CompressionLevel::kNoCompression) {
+      MCTrajectory newMCTrajectory(mcTrajectory.second);
+      _newMCTrajectories->insert(std::pair<art::Ptr<SimParticle>, mu2e::MCTrajectory>(mcTrajectory.second.sim(), newMCTrajectory));
```
Done in commit 08d7eb7

AndrewEdmonds11 commented:
Hi Everyone,

I think I've implemented all the requested changes. Let me know if there is anything I've missed. I will update my slides from last week so that we have an accurate picture of what has been implemented.

Cheers,
Andy

brownd1978 commented on Compression/src/CompressDetStepMCs_module.cc:
```diff
@@ -259,29 +261,31 @@ void mu2e::CompressDetStepMCs::produce(art::Event & event)
   event.put(std::move(_newSimParticles));
   event.put(std::move(_newGenParticles));
-  event.put(std::move(_newMCTrajectories));
+  if (_conf.mcTrajectoryTag() != "") { event.put(std::move(_newMCTrajectories)); }
```
Making the output collections predicate on the input being present will break downstream workflows that depend on a consistent set of products.  I know this change was made in response to a reviewer request, but I ask that you back out this logic, so that the outputs are always created, even if there are no corresponding inputs.  The logic on the inputs can stay.

AndrewEdmonds11 replied:
That makes sense. I've made these changes in commit e8e0ca6

gaponenko replied:
> Making the output collections predicate on the input being present will break downstream workflows that depend on a consistent set of products.

Of course it won't, because in the workflow you have in mind the input
tags will never be empty.  And for a workflow where a product is
missing, it makes zero sense to create it as a "compression" step.
The "ideal" modular solution would not create stuff out of nothing.

I am giving up here and will simply accept this PR, no more requests.

ryuwd commented:
comment to wake FNALbuild up

brownd1978 commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for e8e0ca62ea8b0ffdea0cd590ac2f23127fecef13: build  (Build queue has 4 jobs)

FNALbuild commented:
:sunny: The tests passed at ref e8e0ca62ea8b0ffdea0cd590ac2f23127fecef13.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| merge | :heavy_check_mark: | Merged e8e0ca62ea8b0ffdea0cd590ac2f23127fecef13 at b99a969c52bd5ea6044de8733f38d0263db99447 |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//381/artifact/scons.log). Build time: 15 min 35 sec |
| ceSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//381/artifact/ceSimReco.log) |
| transportOnly | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//381/artifact/transportOnly.log) |
| PS | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//381/artifact/PS.log) |
| g4study | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//381/artifact/g4study.log) |
| cosmicSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//381/artifact/cosmicSimReco.log) |
| rootOverlaps | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//381/artifact/rootOverlaps.log) |
| g4surfaceCheck | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//381/artifact/g4surfaceCheck.log) |
| g4test_03MT | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//381/artifact/g4test_03MT.log) |
| FIXME, TODO count | :heavy_check_mark: | [TODO (0) FIXME (0) in 13 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//381/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [1 errors 0 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//381/artifact/clang-tidy.log) |


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//381/console).
Build artefacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.


Logfiles may also be accessed [at this link.](https://gist.github.com/289421bb41e3acc93e891b0e325a3723)
